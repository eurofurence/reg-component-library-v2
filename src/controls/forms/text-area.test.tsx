/** @jsxImportSource @emotion/react */

import { ChangeEventHandler } from "react"
import { render } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import * as fc from "fast-check"

import TextArea from "./text-area"

describe("TextArea", () => {
	it("renders correctly", () => {
		fc.assert(
			fc.property(fc.string(), fc.string(), (label, placeholder) => {
				const { container } = render(
					<TextArea name="comments" label={label} placeholder={placeholder} />
				)

				expect(container.firstChild).toMatchSnapshot()
			}),
			{ numRuns: 20, seed: 2562346 }
		)
	})

	it("has a label", () => {
		const { queryByLabelText } = render(
			<TextArea name="comments" label="Comments" placeholder="" />
		)

		expect(queryByLabelText("Comments")).toBeTruthy()
	})

	it("enters focus when clicked directly", async () => {
		const { getByRole } = render(
			<TextArea name="comments" label="Comments" placeholder="" />
		)

		const textArea = getByRole("textbox")

		expect(textArea).not.toHaveFocus()

		await userEvent.click(textArea)

		expect(textArea).toHaveFocus()
	})

	it("enters focus when clicked through label", async () => {
		const { getByRole, getByText } = render(
			<TextArea name="comments" label="Comments" placeholder="" />
		)

		const label = getByText("Comments")
		const textArea = getByRole("textbox")

		expect(textArea).not.toHaveFocus()

		await userEvent.click(label)

		expect(textArea).toHaveFocus()
	})

	it.skip("can handle unicode input", async () => {
		await fc.assert(
			fc.asyncProperty(
				fc
					.string({ minLength: 1 })
					.filter(
						(s) =>
							s.trim().length > 0 &&
							!/[!"\\#$%&'()*+,\-./:;<=>?@[\]^_`{|}~]/u.test(s) &&
							!s.includes("__proto__") &&
							!s.includes("arguments")
					),
				async (input) => {
					const onChange = jest.fn() as jest.MockedFunction<
						ChangeEventHandler<HTMLTextAreaElement>
					>
					const { getByRole } = render(
						<TextArea
							name="comments"
							label="Comments"
							placeholder=""
							onChange={onChange}
						/>
					)

					const textArea = getByRole("textbox")

					await userEvent.type(textArea, input)

					expect(
						onChange.mock.calls[onChange.mock.calls.length - 1][0].target.value
					).toEqual(input)
				}
			)
		)
	}, 30000)

	it.skip("can display unicode", async () => {
		await fc.assert(
			fc.asyncProperty(
				fc
					.string({ minLength: 1 })
					.filter(
						(s) =>
							s.trim().length > 0 &&
							!/[!"\\#$%&'()*+,\-./:;<=>?@[\]^_`{|}~]/u.test(s) &&
							!s.includes("__proto__") &&
							!s.includes("arguments")
					),
				async (input) => {
					const onChange = jest.fn() as jest.MockedFunction<
						ChangeEventHandler<HTMLTextAreaElement>
					>
					const { getByRole } = render(
						<TextArea
							name="comments"
							label="Comments"
							placeholder=""
							value={input}
							onChange={onChange}
						/>
					)

					const textArea = getByRole("textbox")

					await userEvent.type(textArea, input)

					expect(textArea).toHaveValue(input)
				}
			)
		)
	}, 30000)

	describe("when controlled", () => {
		it("reflects the input value at all times", async () => {
			const onChange = jest.fn() as jest.MockedFunction<
				ChangeEventHandler<HTMLTextAreaElement>
			>
			const { getByRole } = render(
				<TextArea
					name="comments"
					label="Comments"
					value="test"
					placeholder=""
					onChange={onChange}
				/>
			)

			const textArea = getByRole("textbox")

			expect(textArea).toHaveValue("test")

			await userEvent.type(textArea, "I will be bringing a fursuit.")

			expect(textArea).toHaveValue("test")
		})

		it("raises onChange events when typed in", async () => {
			const onChange = jest.fn() as jest.MockedFunction<
				ChangeEventHandler<HTMLTextAreaElement>
			>
			const { getByRole } = render(
				<TextArea
					name="comments"
					label="Comments"
					value="test"
					placeholder=""
					onChange={onChange}
				/>
			)

			const textArea = getByRole("textbox")

			await userEvent.type(textArea, "I will be bringing a fursuit.")

			expect(onChange).toHaveBeenCalled()
		})

		it("supports readOnly fields", async () => {
			const { getByRole } = render(
				<TextArea
					name="comments"
					label="Comments"
					value="test"
					placeholder=""
					readOnly
				/>
			)

			const textArea = getByRole("textbox")

			await userEvent.type(textArea, "I will be bringing a fursuit.")

			expect(textArea).toHaveValue("test")
		})
	})

	describe("when uncontrolled", () => {
		it("reflects the typed value", async () => {
			const { getByRole } = render(
				<TextArea name="comments" label="Comments" placeholder="" />
			)

			const textArea = getByRole("textbox")

			await userEvent.type(textArea, "I will be bringing a fursuit.")

			expect(textArea).toHaveValue("I will be bringing a fursuit.")
		})

		it("reflects the defaultValue value originally", () => {
			const { getByRole } = render(
				<TextArea
					name="comments"
					label="Comments"
					placeholder=""
					defaultValue="uwu"
				/>
			)

			expect(getByRole("textbox")).toHaveValue("uwu")
		})

		it("raises onChange events when typed in", async () => {
			const onChange = jest.fn() as jest.MockedFunction<
				ChangeEventHandler<HTMLTextAreaElement>
			>
			const { getByRole } = render(
				<TextArea
					name="comments"
					label="Comments"
					placeholder=""
					onChange={onChange}
				/>
			)

			const textArea = getByRole("textbox")

			await userEvent.type(textArea, "I will be bringing a fursuit.")

			expect(onChange).toHaveBeenCalled()
		})
	})
})
