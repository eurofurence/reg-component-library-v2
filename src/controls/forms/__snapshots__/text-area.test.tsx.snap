// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea renders correctly 1`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  />
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="\\\\>"
  />
</label>
`;

exports[`TextArea renders correctly 2`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  >
    !3%^
  </div>
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="Pr"
  />
</label>
`;

exports[`TextArea renders correctly 3`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  />
  <textarea
    class="emotion-2"
    name="comments"
    placeholder=""
  />
</label>
`;

exports[`TextArea renders correctly 4`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  />
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="d"
  />
</label>
`;

exports[`TextArea renders correctly 5`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  >
    {
  </div>
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="uei-"
  />
</label>
`;

exports[`TextArea renders correctly 6`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  >
    &gt;n#C
  </div>
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="B"
  />
</label>
`;

exports[`TextArea renders correctly 7`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  />
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="valueOf"
  />
</label>
`;

exports[`TextArea renders correctly 8`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  />
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="l"
  />
</label>
`;

exports[`TextArea renders correctly 9`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  />
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="3&S@nV0"
  />
</label>
`;

exports[`TextArea renders correctly 10`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  >
    "
  </div>
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="/NvX*&U"
  />
</label>
`;

exports[`TextArea renders correctly 11`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  />
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="w*<#LX\`C"
  />
</label>
`;

exports[`TextArea renders correctly 12`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  >
    __proto__
  </div>
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="\\\\?I'#KKd/"
  />
</label>
`;

exports[`TextArea renders correctly 13`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  >
    V
  </div>
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="|PRHx"
  />
</label>
`;

exports[`TextArea renders correctly 14`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  >
    argumentsr
  </div>
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="zy"
  />
</label>
`;

exports[`TextArea renders correctly 15`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  >
    qdLlwBSkW.
  </div>
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="B4yu"
  />
</label>
`;

exports[`TextArea renders correctly 16`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  >
    |h}xv&%*DB
  </div>
  <textarea
    class="emotion-2"
    name="comments"
    placeholder=""
  />
</label>
`;

exports[`TextArea renders correctly 17`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  >
    $}
  </div>
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="8w\`,8"
  />
</label>
`;

exports[`TextArea renders correctly 18`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  >
    i3-TjDKJ
  </div>
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="sq R&Juds"
  />
</label>
`;

exports[`TextArea renders correctly 19`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  >
    0f
  </div>
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="~"
  />
</label>
`;

exports[`TextArea renders correctly 20`] = `
@media not all and (max-width: 449.99px) {
  form>.emotion-0 {
    grid-column: span 10;
  }
}

.emotion-1 {
  display: block;
  font-weight: 700;
  color: var(--color-grays-500);
  margin-bottom: 0.5em;
}

.emotion-2 {
  min-width: 100%;
  max-width: 100%;
  height: 6.25em;
  border: 2px solid var(--color-grays-300);
  border-radius: 0.1875em;
  padding: 0.75em 1em 1em 1em;
  color: var(--color-grays-900);
}

.emotion-2::-webkit-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::-moz-placeholder {
  color: var(--color-grays-300);
}

.emotion-2:-ms-input-placeholder {
  color: var(--color-grays-300);
}

.emotion-2::placeholder {
  color: var(--color-grays-300);
}

.emotion-2:invalid {
  border-color: var(--color-semantic-error);
}

<label
  class="emotion-0"
>
  <div
    class="emotion-1"
  >
    d5l+N 8e&-
  </div>
  <textarea
    class="emotion-2"
    name="comments"
    placeholder="?}&<?QJ1"
  />
</label>
`;
